### ISSUE TITLE
Crash on deleting a remote branch in combination with filtering

### ISSUE THREAD

**Issue by johan-wiskerke** [jesseduffield/lazygit#3056]
**Describe the bug**

Lazygit crashed/ panicked with a invalid memory address or nil pointer dereference error, after deleting a remote branch.

**To Reproduce**

- Create a remote branch with a name 'zzz' (a name that alphabetically will be the last one in the list.)
- In lazygit go to the `Remotes`, and press enter to show the list of all remote branches. 
- Use the filter `/zzz` and search for the remote 
- Delete the branch (press 'd' and enter.)

lazygit crashes with: 
```
goroutine 1182 [running]:
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x10 pc=0x9c8b0b]

goroutine 827 [running]:
github.com/jesseduffield/lazygit/pkg/commands/models.(*RemoteBranch).FullName(...)
        /home/runner/work/lazygit/lazygit/pkg/commands/models/remote_branch.go:10
github.com/jesseduffield/lazygit/pkg/gui/presentation.GetRemoteBranchListDisplayStrings.func1(0x0, 0x1ab0a5b0501?)
        /home/runner/work/lazygit/lazygit/pkg/gui/presentation/remote_branches.go:12 +0x2b
github.com/samber/lo.Map[...]({0xc000154300?, 0x10, 0xc0000c4b00}, 0xc000a47c50?)
        /home/runner/work/lazygit/lazygit/vendor/github.com/samber/lo/slice.go:29 +0x68
github.com/jesseduffield/lazygit/pkg/gui/presentation.GetRemoteBranchListDisplayStrings({0xc000154300?, 0xc000808200?, 0xc0000c4b00?}, {0x0?, 0x9e0a8a?})
        /home/runner/work/lazygit/lazygit/pkg/gui/presentation/remote_branches.go:11 +0x5d
github.com/jesseduffield/lazygit/pkg/gui/context.NewRemoteBranchesContext.func3(0xc0003adc80?, 0xc000906100?)
        /home/runner/work/lazygit/lazygit/pkg/gui/context/remote_branches_context.go:31 +0x72
github.com/jesseduffield/lazygit/pkg/gui/context.(*ListContextTrait).HandleRender(0xc0003add80)
        /home/runner/work/lazygit/lazygit/pkg/gui/context/list_context_trait.go:90 +0x8d
github.com/jesseduffield/lazygit/pkg/gui.(*Gui).postRefreshUpdate(0xc0002bab00, {0xdef710, 0xc00049c858})
        /home/runner/work/lazygit/lazygit/pkg/gui/view_helpers.go:139 +0xd8
github.com/jesseduffield/lazygit/pkg/gui.(*guiCommon).PostRefreshUpdate(0xc00038ed40?, {0xdef710?, 0xc00049c858?})
        /home/runner/work/lazygit/lazygit/pkg/gui/gui_common.go:32 +0x26
github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshRemotes(0xc00009a480)
        /home/runner/work/lazygit/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:622 +0x1e5
github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func2.8()
        /home/runner/work/lazygit/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:153 +0x1d
github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func2.1.2()
        /home/runner/work/lazygit/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:108 +0xc5
github.com/jesseduffield/lazygit/pkg/utils.Safe.func1()
        /home/runner/work/lazygit/lazygit/pkg/utils/utils.go:97 +0x1b
github.com/jesseduffield/lazygit/pkg/utils.SafeWithError(0x421685?)
        /home/runner/work/lazygit/lazygit/pkg/utils/utils.go:108 +0x67
github.com/jesseduffield/lazygit/pkg/utils.Safe(0xc000b0d160?)
        /home/runner/work/lazygit/lazygit/pkg/utils/utils.go:97 +0x35
created by github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func2.1
        /home/runner/work/lazygit/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:105 +0x156
```

The branch was successfully deleted, but the crash was not expected. 

My conjecture is that the filter uses an index in the list of remote branches. It showed the last branch from the list, and deleting that resulted in filter trying to use an index that no longer is in the list of remote branches.

**Expected behavior**

I expected lazygit to continue, and probably redo the filter operation to look what are the most likely candidates.

**Screenshots**

**Version info:**
_Run `lazygit --version` and paste the result here_
commit=5e388e21c8ca6aa883dbcbe45c47f6fdd5116815, build date=2023-08-07T14:05:48Z, build source=binaryRelease, version=0.40.2, os=windows, arch=amd64, git version=2.41.0.windows.3

_Run `git --version` and paste the result here_
git version 2.41.0.windows.3

**Additional context**

**Note:** please try updating to the latest version or [manually building](https://github.com/jesseduffield/lazygit/#manual) the latest `master` to see if the issue still occurs.

**Comment by AzraelSec** [comment]
Thank you, @johan-wiskerke, for reporting this. I can reproduce the issue, and I'm able to trigger it also for the remote tags:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x9bd26e]

goroutine 642 [running]:
github.com/jesseduffield/lazygit/pkg/gui/context.NewTagsContext.func3.GetTagListDisplayStrings.func1(0xc000088050?, 0xc000774358?)
        /home/azraelsec/dev/lazygit/pkg/gui/presentation/tags.go:13 +0xe
github.com/samber/lo.Map[...]({0xc000088080?, 0x1, 0x1}, 0xc00084bd00?)
        /home/azraelsec/dev/lazygit/vendor/github.com/samber/lo/slice.go:29 +0x74
github.com/jesseduffield/lazygit/pkg/gui/presentation.GetTagListDisplayStrings(...)
        /home/azraelsec/dev/lazygit/pkg/gui/presentation/tags.go:12
github.com/jesseduffield/lazygit/pkg/gui/context.NewTagsContext.func3(0xc00084bd70?, 0x9bf6e6?)
        /home/azraelsec/dev/lazygit/pkg/gui/context/tags_context.go:30 +0xa5
github.com/jesseduffield/lazygit/pkg/gui/context.(*ListRenderer).renderLines(0xc000054710, 0xffffffffffffffff, 0xffffffffffffffff)
        /home/azraelsec/dev/lazygit/pkg/gui/context/list_renderer.go:88 +0x1b1
github.com/jesseduffield/lazygit/pkg/gui/context.(*ListContextTrait).HandleRender(0xc000054700)
        /home/azraelsec/dev/lazygit/pkg/gui/context/list_context_trait.go:88 +0x45
github.com/jesseduffield/lazygit/pkg/gui.(*Gui).postRefreshUpdate(0xc000436000, {0xdd38a8, 0xc000408510})
        /home/azraelsec/dev/lazygit/pkg/gui/view_helpers.go:139 +0xc7
github.com/jesseduffield/lazygit/pkg/gui.(*guiCommon).PostRefreshUpdate(0xc000438408?, {0xdd38a8?, 0xc000408510?})
        /home/azraelsec/dev/lazygit/pkg/gui/gui_common.go:32 +0x25
github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).refreshTags(0xc0002799c0)
        /home/azraelsec/dev/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:406 +0xf8
github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func2.7()
        /home/azraelsec/dev/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:153 +0x17
github.com/jesseduffield/lazygit/pkg/gui/controllers/helpers.(*RefreshHelper).Refresh.func2.1.1({0x0?, 0x0?})
        /home/azraelsec/dev/lazygit/pkg/gui/controllers/helpers/refresh_helper.go:105 +0x13
github.com/jesseduffield/gocui.(*Gui).onWorkerAux(0x810cc0?, 0xc00063d0e0?, {0xdc7ef8?, 0xc000984140?})
        /home/azraelsec/dev/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:680 +0x66
github.com/jesseduffield/gocui.(*Gui).OnWorker.func1()
        /home/azraelsec/dev/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:667 +0x2b
created by github.com/jesseduffield/gocui.(*Gui).OnWorker in goroutine 638
        /home/azraelsec/dev/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:666 +0x92
exit status 2
```

The issue is caused by the fact that after changing the list of branches for the remote origin, the related [FilteredListViewModel](https://github.com/AzraelSec/lazygit/blob/c39fafe6ec6fb64d1510adef8980e93aebb334d5/pkg/gui/context/remote_branches_context.go#L23-L28) does not re-apply the current filter, and the GetFilteredList method [returns nil](https://github.com/AzraelSec/lazygit/blob/c39fafe6ec6fb64d1510adef8980e93aebb334d5/pkg/gui/context/filtered_list.go#L52) as part of the result. The [filteredIndices](https://github.com/AzraelSec/lazygit/blob/c39fafe6ec6fb64d1510adef8980e93aebb334d5/pkg/gui/context/filtered_list.go#L13) field is out-of-date after removing an element.
I can fix it by applying the filter on the list every time the [GetFilteredList](https://github.com/jesseduffield/lazygit/compare/master...AzraelSec:lazygit:fix-filtered-list-element-removal) is invoked, but it's not the right way. We should invalidate the list every time a refresh occurs or change the filtered list view model logic to apply the filter on-read. What are your thoughts [@jesseduffield](https://github.com/jesseduffield) [@stefanhaller](https://github.com/stefanhaller)? Happy to work on this.

**Comment by stefanhaller** [comment]
Related to #2816.

@AzraelSec Great analysis! I agree that it would be best to reapply the filter on every refresh. I made a draft PR that does this, it seems to fix the problem for me: #3058

**Comment by AzraelSec** [comment]
Thank you, @stefanhaller! As said in the PR, I was honestly unsure how to write this patch. I've tested your code locally and left just a couple of suggestions. Happy to have helped! :)

**Comment by stefanhaller** [comment]
Fixed by #3058.
